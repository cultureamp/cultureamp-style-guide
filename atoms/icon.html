<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
  <title>Kaizen :: Icon</title>
  
  <link rel="stylesheet" href="/cultureamp-style-guide/assets/main-cebf848ed1fb02c3cc94.bundle.css">
  <script src="/cultureamp-style-guide/assets/main-cebf848ed1fb02c3cc94.bundle.js"></script>
</head>

  <body>
    <nav id="guide-nav">
  <ul>
    <li><a href="/cultureamp-style-guide/">home</a></li>
    

<li><a href="/cultureamp-style-guide/visuals/">Visuals</a>
  
</li>

    

<li><a href="/cultureamp-style-guide/atoms/">Atoms</a>
  
  <ul>
    
      <li><a href="/cultureamp-style-guide/atoms/button">Button</a></li>
    
      <li><a href="/cultureamp-style-guide/atoms/icon">Icon</a></li>
    
  </ul>
  
</li>

    

<li><a href="/cultureamp-style-guide/molecules/">Molecules</a>
  
</li>

    

<li><a href="/cultureamp-style-guide/organisms/">Organisms</a>
  
</li>

    

<li><a href="/cultureamp-style-guide/templates/">Templates</a>
  
</li>

    

<li><a href="/cultureamp-style-guide/pages/">Pages</a>
  
</li>

  </ul>
</nav>

    <h1>Icon</h1>
    <div class="componentDescription">
      <p>An icon! To control its color, set <code class="highlighter-rouge">color</code> on a parent element, and it will be
inherited.</p>

    </div>
    <div data-react-component="Demo"
  data-react-props='{
    "component": "IconDemo",
    "presets": [{"name":"Meaningful","props":{"icon":"accents-warning","role":"img","title":"Warning"}},{"name":"Presentational","props":{"icon":"accents-warning","role":"presentation"}},{"name":"Inherit size","props":{"icon":"accents-warning","inheritSize":true,"role":"presentation"}}]
  }'></div>

  </body>
</html>
