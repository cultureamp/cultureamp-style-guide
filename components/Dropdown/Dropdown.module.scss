@import '~cultureamp-style-guide/styles/color';
@import '~cultureamp-style-guide/styles/border';
@import '~cultureamp-style-guide/styles/layers';

$width: 248px;

.fieldDropdownIconInner {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  position: relative;
}

.dropdownIcon {
  composes: buttonReset from 'components/HtmlContent.module.scss';
  cursor: pointer;
  display: flex;
  flex: 1;
  height: 100%;
  width: 100%;
  align-items: center;
  justify-content: center;
}

.dropdownHoverArea {
  .dropdownIcon {
    opacity: 0;
    &.isOpen,
    &:focus {
      opacity: 1;
    }
  }
  &:hover .dropdownIcon {
    opacity: 1;
  }
}

.menuContainer {
  position: absolute;
  width: $width;
  z-index: $ca-z-index-dropdown;
}
