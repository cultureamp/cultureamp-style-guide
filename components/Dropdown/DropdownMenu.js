// @flow
import PropTypes from 'prop-types';
import * as React from 'react';
import styles from './Dropdown.module.scss';

type DropdownMenuProps = {
  children: React.Node,
  hideDropdownMenu: () => void,
  position: ?{
    top: number,
    bottom: number,
    left: number,
    right: number,
  },
};

export default class DropdownMenu extends React.Component<DropdownMenuProps> {
  static displayName = 'DropdownMenu';
  menu: ?HTMLDivElement;

  componentDidMount() {
    document.addEventListener('click', this.handleDocumentClick, false);
    window.addEventListener('resize', this.handleDocumentResize, false);
    this.positionMenu();
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handleDocumentClick, false);
    window.removeEventListener('resize', this.handleDocumentResize, false);
  }

  positionMenu() {
    const menu = this.menu;
    
    if (!this.props.position || !menu) {
      return;
    }
    
    const pos = this.props.position;
    const heightBetweenTopOfRowAndDropdownIcon = (pos.bottom - pos.top) / 2;
    const { innerHeight } = window;
    const rect = menu.getBoundingClientRect();

    if (pos.bottom > innerHeight - rect.height) {
      menu.style.bottom = '24px';
      menu.style.top = 'auto';
    } else {
      menu.style.top = '24px';
      menu.style.bottom = 'auto';
    }
  }

  handleDocumentClick = (e: MouseEvent) => {
    if (
      this.menu &&
      e.target instanceof Node &&
      !this.menu.contains(e.target)
    ) {
      this.props.hideDropdownMenu();
    }
  };

  handleDocumentResize = () => {
    this.props.hideDropdownMenu();
  };

  render() {
    const props = this.props;
    return (
      <div
        className={styles.menuContainer}
        ref={m => (this.menu = m)}
        onClick={() => props.hideDropdownMenu()}
      >
        {props.children}
      </div>
    );
  }
}
