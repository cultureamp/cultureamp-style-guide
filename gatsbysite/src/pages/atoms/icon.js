import React from 'react';
import Link from 'gatsby-link';
import Demo from 'components/Demo';
import Icon from '../../../../components/Icon/Icon.js';
import configureIcon from 'cultureamp-style-guide/icons/configure.svg';

const iconPresets = [
  {
    name: 'Meaningful',
    props: { icon: configureIcon, role: 'img', title: 'Warning' },
  },
  {
    name: 'Presentational',
    props: { icon: configureIcon, role: 'presentation' },
  },
  {
    name: 'Inherit size',
    props: { icon: configureIcon, inheritSize: true, role: 'presentation' },
  },
];

const Page = () => (
  <div>
    <h1>Icons</h1>
    <p>
      An icon! To control its color, set color on a parent element, and it will
      be inherited.
    </p>
    <Demo component={Icon} presets={iconPresets} />
    <Link to="/">Go back to the homepage</Link>
  </div>
);

export default Page;
