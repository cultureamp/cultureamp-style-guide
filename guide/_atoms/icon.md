---
title: Icon
component: IconDemo
presets:
  - name: Meaningful
    props:
      icon: accents-warning
      role: img
      title: Warning
  - name: Presentational
    props:
      icon: accents-warning
      role: presentation
  - name: Inherit size
    props:
      icon: accents-warning
      inheritSize: true
      role: presentation
---

An icon! To control its color, set `color` on a parent element, and it will be
inherited.
