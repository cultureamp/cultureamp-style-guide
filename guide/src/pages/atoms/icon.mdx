---
imports:
  Demo: components/Demo.js
  Link: components/Link.js
  Icon: ../../../../components/Icon/Icon.js
  iconPresets: ./_iconPresets.js
---

# Icon

An icon! To control its color, set color on a parent element, and it will be inherited.

<Demo component={Icon} presets={iconPresets} />

<Link to="/">Go back to the homepage</Link>