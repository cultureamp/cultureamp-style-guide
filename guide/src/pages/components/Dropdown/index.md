---
imports:
  Demo: components/Demo.js
  InlineNotification: cultureamp-style-guide/components/Notification/InlineNotification.js
  IntroParagraph: components/IntroParagraph.js
  presets: ./_presets.js
---

<InlineNotification persistent={true} type="cautionary" title="Out of date">This documentation is out of date while we update the site. Until then, Culture Amp employees can refer to the <a href="https://cultureamp.atlassian.net/wiki/spaces/CA/pages/916161089/Kaizen+Visual+Documentation+Status">visual documentation status page on the wiki</a>.</InlineNotification>

# Dropdown

<IntroParagraph>

A simple dropdown menu to be used for opening links or triggering actions.
Do not use this if the user is selecting a value, prefer a select box instead.

</IntroParagraph>

<Demo presets={presets} />
