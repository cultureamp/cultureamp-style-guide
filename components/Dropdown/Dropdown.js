// @flow
import PropTypes from 'prop-types';
import * as React from 'react';
import classNames from 'classnames';
import styles from './Dropdown.module.scss';
import Icon from '../Icon/Icon';
import defaultIcon from '../../icons/ellipsis.svg';
import chevronDown from '../../icons/chevron-down.svg';
import DropdownMenu from './DropdownMenu';
import type { SvgAsset } from '../Icon/Icon.js';

type DropdownState = {
  isMenuVisible: boolean,
};

type DropdownProps = {
  icon?: SvgAsset,
  label?: React.Node,
  children: React.Node,
  menuVisible?: boolean,
  controlAction?: boolean,
  automationId?: string,
};

export default class Dropdown extends React.Component<
  DropdownProps,
  DropdownState
> {
  static displayName = 'Dropdown';
  dropdownButton: ?HTMLButtonElement;

  constructor(props: DropdownProps) {
    super(props);
    this.state = {
      isMenuVisible: Boolean(props.menuVisible),
    };
  }

  toggleDropdownMenu = (e: SyntheticEvent<HTMLButtonElement>) => {
    e.stopPropagation();
    const currentState = this.state.isMenuVisible;
    this.setState({
      isMenuVisible: !currentState,
    });
  };

  hideDropdownMenu = () => {
    this.setState({
      isMenuVisible: false,
    });
  };

  getPosition() {
    return this.dropdownButton
      ? this.dropdownButton.getBoundingClientRect()
      : null;
  }

  renderDropdownMenu() {
    return (
      <DropdownMenu
        hideDropdownMenu={this.hideDropdownMenu}
        position={this.getPosition()}
      >
        {this.props.children}
      </DropdownMenu>
    );
  }

  render() {
    let { icon, label, controlAction, automationId } = this.props;
    if (!icon && !label) {
      icon = defaultIcon;
    }
    const btnClass = classNames(styles.dropdownButton, {
      [styles.dropdownControlAction]: controlAction,
      [styles.isOpen]: this.state.isMenuVisible,
    });
    return (
      <div className={styles.dropdown}>
        <button
          className={btnClass}
          onClick={this.toggleDropdownMenu}
          onMouseDown={e => e.preventDefault()}
          ref={k => (this.dropdownButton = k)}
          data-automation-id={automationId}
        >
          {icon && (
            <span className={styles.dropdownIcon}>
              <Icon icon={icon} role="img" title="Open menu" />
            </span>
          )}
          {label && <span className={styles.dropdownLabel}>{label}</span>}
          {label &&
            controlAction && (
              <span className={styles.chevronIcon}>
                <Icon icon={chevronDown} role="img" title="Open menu" />
              </span>
            )}
        </button>
        {this.state.isMenuVisible && this.renderDropdownMenu()}
      </div>
    );
  }
}
