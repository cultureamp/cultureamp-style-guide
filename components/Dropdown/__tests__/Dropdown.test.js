import React from 'react';
import { mount } from 'enzyme';
import Dropdown from '../Dropdown';
import icon from '../../../icons/add.svg';

describe('Dropdown', () => {
  test('renders default view', () => {
    const component = mount(<Dropdown />);
    expect(component).toMatchSnapshot();
  });

  test('renders drop down with icon', () => {
    const component = mount(<Dropdown icon={icon} />);
    expect(component).toMatchSnapshot();
  });

  test('renders drop down with icon and label', () => {
    const component = mount(<Dropdown icon={icon} label="add" />);
    expect(component).toMatchSnapshot();
  });

  test('renders drop down with only label', () => {
    const component = mount(<Dropdown label="add" />);
    expect(component).toMatchSnapshot();
  });

  test('renders control action dropdown', () => {
    const component = mount(<Dropdown icon={icon} label="add" controlAction />);
    expect(component).toMatchSnapshot();
  });

  test('renders reversed color control action dropdown', () => {
    const component = mount(
      <Dropdown icon={icon} label="add" controlAction reversedColor />
    );
    expect(component).toMatchSnapshot();
  });

  test('shows dropdown menu when clicking on the button', () => {
    const component = mount(<Dropdown />);

    component.find('.dropdownButton').simulate('click');

    expect(component.find('DropdownMenu').exists()).toEqual(true);
  });
});
