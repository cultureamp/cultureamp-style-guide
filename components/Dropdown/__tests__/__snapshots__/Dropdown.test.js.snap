// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders control action dropdown 1`] = `
<Dropdown
  controlAction={true}
  icon={null}
  iconPosition="start"
  label="add"
>
  <div
    className="dropdown"
  >
    <button
      className="dropdownButton dropdownControlAction"
      onClick={[Function]}
      onMouseDown={[Function]}
    >
      <span
        className="dropdownIcon"
      >
        <Icon
          icon={null}
          inheritSize={false}
          role="img"
          title="Open menu"
        >
          <svg
            aria-labelledby="id-3-title"
            className="icon"
            focusable="false"
            role="img"
            viewBox="viewBox"
          >
            <title
              id="id-3-title"
            >
              Open menu
            </title>
            <use
              xlinkHref="#id"
            />
          </svg>
        </Icon>
      </span>
      <span
        className="dropdownLabel"
      >
        add
      </span>
      <span
        className="chevronIcon"
      >
        <Icon
          icon={null}
          inheritSize={false}
          role="img"
          title="Open menu"
        >
          <svg
            aria-labelledby="id-4-title"
            className="icon"
            focusable="false"
            role="img"
            viewBox="viewBox"
          >
            <title
              id="id-4-title"
            >
              Open menu
            </title>
            <use
              xlinkHref="#id"
            />
          </svg>
        </Icon>
      </span>
    </button>
  </div>
</Dropdown>
`;

exports[`Dropdown renders default view 1`] = `
<Dropdown
  iconPosition="start"
>
  <div
    className="dropdown"
  >
    <button
      className="dropdownButton"
      onClick={[Function]}
      onMouseDown={[Function]}
    />
  </div>
</Dropdown>
`;

exports[`Dropdown renders drop down with icon 1`] = `
<Dropdown
  icon={null}
  iconPosition="start"
>
  <div
    className="dropdown"
  >
    <button
      className="dropdownButton"
      onClick={[Function]}
      onMouseDown={[Function]}
    >
      <span
        className="dropdownIcon"
      >
        <Icon
          icon={null}
          inheritSize={false}
          role="img"
          title="Open menu"
        >
          <svg
            aria-labelledby="id-1-title"
            className="icon"
            focusable="false"
            role="img"
            viewBox="viewBox"
          >
            <title
              id="id-1-title"
            >
              Open menu
            </title>
            <use
              xlinkHref="#id"
            />
          </svg>
        </Icon>
      </span>
    </button>
  </div>
</Dropdown>
`;

exports[`Dropdown renders drop down with icon and label 1`] = `
<Dropdown
  icon={null}
  iconPosition="start"
  label="add"
>
  <div
    className="dropdown"
  >
    <button
      className="dropdownButton"
      onClick={[Function]}
      onMouseDown={[Function]}
    >
      <span
        className="dropdownIcon"
      >
        <Icon
          icon={null}
          inheritSize={false}
          role="img"
          title="Open menu"
        >
          <svg
            aria-labelledby="id-2-title"
            className="icon"
            focusable="false"
            role="img"
            viewBox="viewBox"
          >
            <title
              id="id-2-title"
            >
              Open menu
            </title>
            <use
              xlinkHref="#id"
            />
          </svg>
        </Icon>
      </span>
      <span
        className="dropdownLabel"
      >
        add
      </span>
    </button>
  </div>
</Dropdown>
`;

exports[`Dropdown renders drop down with only label 1`] = `
<Dropdown
  iconPosition="start"
  label="add"
>
  <div
    className="dropdown"
  >
    <button
      className="dropdownButton"
      onClick={[Function]}
      onMouseDown={[Function]}
    >
      <span
        className="dropdownLabel"
      >
        add
      </span>
    </button>
  </div>
</Dropdown>
`;

exports[`Dropdown renders reversed control action dropdown 1`] = `
<Dropdown
  controlAction={true}
  icon={null}
  iconPosition="start"
  label="add"
  reversed={true}
>
  <div
    className="dropdown"
  >
    <button
      className="dropdownButton dropdownControlAction reversed"
      onClick={[Function]}
      onMouseDown={[Function]}
    >
      <span
        className="dropdownIcon"
      >
        <Icon
          icon={null}
          inheritSize={false}
          role="img"
          title="Open menu"
        >
          <svg
            aria-labelledby="id-5-title"
            className="icon"
            focusable="false"
            role="img"
            viewBox="viewBox"
          >
            <title
              id="id-5-title"
            >
              Open menu
            </title>
            <use
              xlinkHref="#id"
            />
          </svg>
        </Icon>
      </span>
      <span
        className="dropdownLabel"
      >
        add
      </span>
      <span
        className="chevronIcon"
      >
        <Icon
          icon={null}
          inheritSize={false}
          role="img"
          title="Open menu"
        >
          <svg
            aria-labelledby="id-6-title"
            className="icon"
            focusable="false"
            role="img"
            viewBox="viewBox"
          >
            <title
              id="id-6-title"
            >
              Open menu
            </title>
            <use
              xlinkHref="#id"
            />
          </svg>
        </Icon>
      </span>
    </button>
  </div>
</Dropdown>
`;
